### ISSUE TITLE
BaceConfig class shallow copy method

### ISSUE THREAD

**Issue by isaac-tfn** [Transport-for-the-North/caf.toolkit#129]
The .copy method for BaseConfig classes is shallow by default, so copying an instance then altering that copy also alters the original. I have only experienced this with a list, I don't know if other types will behave in the same way. This also may be the intended and correct way for the copy method to act.

`new_conf = self.input.copy()
new_conf.naming_order[0] = 'example_name'`

Where self.input is a class inheriting from toolkit's BaseConfig, and naming_order is a list of strings. The above code will alter naming order in the original class instance as well as the copy. Using copy.deepcopy fixes this issue.

`new_conf = deepcopy(self.input)
new_conf.naming_order[0] = 'example_name'`

If this is not a bug or issue and is just me not understanding copy feel free to remove as an issue!

**Comment by MattBuckley-TfN** [comment]
`BaseConfig` doesn't define any copy functionality, it is all defined in the parent class [`pydantic.BaseModel`](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel). The `BaseModel` class defines the `__copy__` and `__deepcopy__` dunder methods so that `copy.copy` and `copy.deepcopy` work.

The `BaseModel.copy` method is shallow by default (which I think is pretty standard) but has the deep parameter which can be set to true `.copy(deep=True)`, the `copy` method on `BaseModel` is deprecated and `BaseModel.model_copy` should be used instead, which seems to work in the same way.

We could potentially write a copy method for `BaseConfig` which was just a wrapper around `model_copy` if it would avoid confusion but I think we would probably still want to default to `deep=False` as this seems to be the standard.

**Comment by isaac-tfn** [comment]
If the 'copy' method is deprecated/model_copy should always be used then might be worth adding our own copy as a wrapper yes, otherwise probably leave as is

**Comment by MattBuckley-TfN** [comment]
> If the 'copy' method is deprecated/model_copy should always be used then might be worth adding our own copy as a wrapper yes, otherwise probably leave as is

I agree, a simple copy method which just wraps `model_copy`. Although this probably isn't a high priority as for now `model_copy` can be used
